File: Diagrammatize.Design.Pattern/code/cxx/Memento/A4/game/Gamer.cpp
#include "Gamer.hpp"
#include <iostream>
#include <sstream>

#include "starts_with.hpp"

std::vector<std::string> Gamer::fruitsname = {
    "苹果", "葡萄", "香蕉", "橘子",
};

Gamer::Gamer(int money_): money(money_) {
}

int Gamer::getMoney() {
    return money;
}

void Gamer::bet() {
    int dice = random.randomInt(1, 6);
    if (dice == 1) {
        money += 100;
        std::cout << "所持金钱增加了。" << std::endl;
    } else if (dice == 2) {
        money /= 2;
        std::cout << "所持金钱减半了。" << std::endl;
    } else if (dice == 6) {
        std::string f = getFruit();
        std::cout << "获得了水果(" << f << ")。" << std::endl;
        fruits.push_back(f);
    } else {
        std::cout << "什么都没有发生。" << std::endl;
    }
}

Memento* Gamer::createMemento() {
    Memento* m = new Memento(money);
    for (auto& f: fruits) {
        if (starts_with(f, "好吃的")) {
            m->addFruit(f);
        }
    }
    return m;
}

void Gamer::restoreMemento(Memento* memento) {
    money = memento->getMoney();
    fruits = memento->getFruits();
}

std::string Gamer::toString() const {
    std::ostringstream os;
    os << "[money = " << money 
        << ", fruits = [ ";
    bool first = true;
    for (auto& f: fruits) {
        if (first) {
            first = false;
        } else {
            os << ", ";
        }
        os << f;
    }
    os << " ]]";
    return os.str();
}

std::string Gamer::getFruit() {
    std::string prefix = "";
    if (random.randomInt(0, 1)) {
        prefix = "好吃的";
    }
    return prefix + fruitsname[random.randomInt(0, fruitsname.size()-1)];
}

File: Diagrammatize.Design.Patterns/code/python/Memento/A4/game/Gamer.py
import time
import random

from .Memento import Memento

class Gamer:
    fruitsname = [ "苹果", "葡萄", "香蕉", "橘子", ]

    def __init__(self, money):
        self.money = money
        self.fruits = list()
        self.random = random.Random(time.time())

    def getMoney(self):
        return self.money

    def bet(self):
        dice = self.random.randint(1, 6)
        if dice == 1:
            self.money += 100
            print("所持金钱增加了。")
        elif dice == 2:
            self.money //= 2
            print("所持金钱减半了。")
        elif dice == 6:
            f = self.getFruit()
            print("获得了水果({})。".format(f));
            self.fruits.append(f)
        else:
            print("什么都没有发生。")

    def createMemento(self):
        m = Memento(self.money)
        for f in self.fruits:
            if f.startswith("好吃的"):
                m.addFruit(f)

        return m

    def restoreMemento(self, memento):
        self.money = memento.getMoney()
        self.fruits = memento.getFruits()

    def __str__(self):
        return "[money = {}, fruits = {}]".format(self.money, self.fruits)

    def getFruit(self):
        prefix = ""
        if self.random.randint(0, 1):
            prefix = "好吃的"

        return prefix + Gamer.fruitsname[self.random.randint(0, len(Gamer.fruitsname)-1)]
